### PR TITLE
Adding call for contribution for the S4 model

### PR DESCRIPTION
**Pull request by Rocketknight1**
Looking for community contributors to add the S4 model! ([repo](https://github.com/HazyResearch/state-spaces), [paper](https://arxiv.org/abs/2111.00396))

### CODE REVIEW

File: templates/adding_a_new_model/open_model_proposals/ADD_S4.md [hunk 1]
```diff
@@ -0,0 +1,1028 @@
...
+**Note:** You don't need to have CUDA installed. Making the new model
+work on CPU is sufficient.
+
+5.  To port *S4*, you will also need access to its
+    original repository:
+
+```bash
+git clone https://github.com/HazyResearch/state-spaces.git 
+cd s4
+pip install -e .
```
Comment (patrickvonplaten) [inline templates/adding_a_new_model/open_model_proposals/ADD_S4.md:286]: Is this enough or might there be other things to install?

File: templates/adding_a_new_model/open_model_proposals/ADD_S4.md [hunk 2]
```diff
@@ -0,0 +1,1028 @@
...
+Transformers implementation of *S4*. It is not advised to
+write the conversion script from scratch, but rather to look through
+already existing conversion scripts in 🤗 Transformers for one that has
+been used to convert a similar model that was written in the same
+framework as *S4*. Usually, it is enough to copy an
+already existing conversion script and slightly adapt it for your use
+case. Don't hesitate to ask [Matt](https://github.com/rocketknight1) to point you to a
+similar already existing conversion script for your model.
+
+-   If you are porting a model from TensorFlow to PyTorch, a good
```
Comment (patrickvonplaten) [inline templates/adding_a_new_model/open_model_proposals/ADD_S4.md:665]: Think this is PyTorch to PyTorch so maybe delete this?

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me! Left some nits

**Comment by kamalkraj** [comment]
Hi @Rocketknight1,
I would like to work on this S4 model

**Comment by Rocketknight1** [comment]
Hi @kamalkraj, sure thing! I thought this was quite an unusual and challenging model, but you've done some really good work in other PRs, so I think you should be able for it. Note that since the original repository is PyTorch-only, we'll probably want a PyTorch version of the model only to start, and we can leave a TF port for a future PR. Is that okay with you?

**Comment by kamalkraj** [comment]
Yes, @Rocketknight.

**Comment by Rocketknight1** [comment]
Sure! In that case I'll leave this PR un-merged for now. You can see most of the details in the .md file. The best place to get started is probably to get their [standalone example](https://github.com/HazyResearch/state-spaces/blob/main/example.py) working with `pykeops`. If you encounter any difficulties, or you're not sure how to proceed, you can let me know here or on Slack. We have direct communication with the original paper authors too, so we should be able to ask them if we encounter any issues with the underlying theory. Thank you for this!

**Comment by kamalkraj** [comment]
Okay

**Comment by kamalkraj** [comment]
@Rocketknight1 
Got the standalone version working with pykeops
https://github.com/kamalkraj/S4-Standalone

**Comment by Rocketknight1** [comment]
@kamalkraj Nice!

**Comment by kamalkraj** [comment]
Hi @Rocketknight1,
I just skimmed through the paper and code, If I understood correctly, I only need to port the autoregressive Language Model to HF, right?

**Comment by Rocketknight1** [comment]
Hi @kamalkraj - for now, that should be fine, yes. I think it's best to start with porting a single model correctly and then we can decide what to do afterwards!

**Comment by kamalkraj** [comment]
Okay

**Comment by kamalkraj** [comment]
I have started working on the model.
make fixup is throwing an error ```  S4Config in TYPE_HINT but not in _import_structure.```

```python
from typing import TYPE_CHECKING

# rely on isort to merge the imports
from ...file_utils import _LazyModule, is_torch_available


_import_structure = {"configuration_s4": ["S4Config"]}


if is_torch_available():
    _import_structure["modeling_s4"] = ["S4ForCausalLM", "S4Layer", "S4Model", "S4PreTrainedModel"]


if TYPE_CHECKING:
    from .configuration_s4 import S4Config

    if is_torch_available():
        from .modeling_s4 import S4ForCausalLM, S4Layer, S4Model, S4PreTrainedModel

else:
    import sys

    sys.modules[__name__] = _LazyModule(__name__, globals()["__file__"], _import_structure)
```

Screenshot 2021-12-14 at 8 54 31 PM

@Rocketknight1 @sgugger

**Comment by sgugger** [comment]
The script is not used to have something taht fits in one line for the declaration of `_import_structure`. Will fix it one day but in the meantime, you should replace it by 
```
# fmt: off
_import_structure = {
    "configuration_s4": ["S4Config"]
}
# fmt: on
```
The flags fmt will tell black to be a good boy and leave the style as is, and the script will parse the init properly.

**Comment by Rocketknight1** [comment]
Closing this call for contributions PR since @kamalkraj is already working on the actual model PR.

**Pull request closed without merging**